src/slack_interaction_handler.ts

```diff
@@ -319,3 +344,201 @@ export async function handleVolunteerUpdate({
+  const userId = MD5.hex(userPhoneNumber);
+
+  const commandType = 'RESET_DEMO';
+  // Ignore Prettier formatting because this object needs to adhere to JSON strigify requirements.
+  // prettier-ignore
+  const modalPrivateMetadata = {
+    "commandType": commandType,
+    "userId": userId,
+    "userPhoneNumber": userPhoneNumber,
+    "twilioPhoneNumber": twilioPhoneNumber,
+    "slackChannelId": payload.channel.id,
+    "slackParentMessageTs": payload.message.ts,
+    "originatingSlackUserName": originatingSlackUserName,
+    "originatingSlackUserId": payload.user.id,
+    "slackChannelName": slackChannelName,
+    "actionTs": payload.action_ts
+  } as SlackModalPrivateMetadata;
+
+  const redisUserInfoKey = `${userId}:${twilioPhoneNumber}`;
+  const userInfo = (await RedisApiUtil.getHash(
+    redisClient,
+    redisUserInfoKey
+  )) as UserInfo;
+
+  if (!userInfo) {
+    modalPrivateMetadata.success = false;
+    modalPrivateMetadata.failureReason = 'no_user_info';
+    await DbApiUtil.logCommandToDb(modalPrivateMetadata);
+    throw new Error(
+      `SLACKINTERACTIONHANDLER.receiveResetDemo: Interaction received for voter who has redisData but not userInfo: active redisData key is ${payload.channel.id}:${payload.message.ts}, userInfo key is ${redisUserInfoKey}.`
+    );
+  }
+
+  let slackView;
+  if (!userInfo.isDemo) {
+    modalPrivateMetadata.success = false;
+    modalPrivateMetadata.failureReason = 'non_demo';
+    await DbApiUtil.logCommandToDb(modalPrivateMetadata);
+    logger.info(
+      `SLACKINTERACTIONHANDLER.receiveResetDemo: Volunteer tried to reset demo on non-demo voter.`
+    );
+    slackView = SlackBlockUtil.getErrorSlackView(
+      'demo_reset_error_not_demo',
+      'This shortcut is strictly for demo conversations only. Please reach out to an admin for assistance.'
+    );
+  } else if (!(payload.channel.id === userInfo.activeChannelId)) {
+    modalPrivateMetadata.success = false;
+    modalPrivateMetadata.failureReason = 'non_active_thread';
+    await DbApiUtil.logCommandToDb(modalPrivateMetadata);
+    logger.info(
+      `SLACKINTERACTIONHANDLER.receiveResetDemo: Volunteer issued reset demo command from #${payload.channel.id} but voter active channel is ${userInfo.activeChannelId}.`
+    );
+    slackView = SlackBlockUtil.getErrorSlackView(
+      'demo_reset_error_not_active_thread',
+      `This voter is no longer active in this thread. Please reach out to the folks at *#${userInfo.activeChannelName}*.`
+    );
+  } else {
+    logger.info(
+      `SLACKINTERACTIONHANDLER.receiveResetDemo: Reset demo command is valid.`
+    );
+
+    // Store the relevant information in the modal so that when the requested action is confirmed
+    // the data needed for the necessary actions is available.
+    slackView = SlackBlockUtil.resetConfirmationSlackView(
+      commandType,
+      modalPrivateMetadata
+    );
+  }
+
+  await SlackApiUtil.renderModal(payload.trigger_id, slackView);
```

Comment:
One potential issue here: The `trigger_id` expires after 3 seconds, and we're doing this work inside a background task after a DB command. You'll get here within 3 seconds _most_ of the time, but if there's a large amount of the traffic / network congestion / etc., the command might fail.

This isn't a part of Slack's API that I particularly love, but the recommended way to handle this is to open the modal ASAP in a loading state and then to update the modal with stuff as needed. This is fine for now (and it looks like it's currently processing under 3s so maybe it's fine?), but I'm working on some utils on my end that'll make the open-then-update flow a little nicer.

Reply:
Great point, I was worried about that and hadn't thought of that solution. Since we're looking to start volunteer practice today and it does appear to be working for now (and if it failed, luckily it would only affect volunteers and not voters), what are your thoughts on submitting this change and following up with a fix?